### DISCUSSION TITLE
Spatially dependent material properies

### DISCUSSION THREAD

**Discussion by MojtabaFarrokh** [thelfer/tfel#452] [General]
Hi
Is there any way in Mfront to define a material whose properties depend on the coordinates X,Y,Z?
FEM packages such as abaqus and calculix pass the integration point coordinates in the user-defined subroutine UMAT as an input parameter.

**Comment by thelfer** [comment]
Dear @MojtabaFarrokh,

Spatial coordinates are currently not accessible  in `MFront`' behaviours for at least two reasons:

- behaviours which explicitly depends on spatial coordinates are not physical (the behaviour must be invariant by change of coordinates). For the same reason, the time is not available (but the time increment is).
- the spatial coordiantes are not available in most solvers.

In Abaqus, you can use external state variables to solve this issue. In calculix, i don't know.

Regards,

Thomas

**Reply by thelfer** [reply]
To be complete, there is another solution: modify the `UMAT` source code generated by `mfront` and compute the quantity you need in the modified `UMAT` and pass those quantities as external state variables. If such solution is fine for you, I can show you how to do it.

**Comment by MojtabaFarrokh** [comment]
Dear @thelfer,
Thank you for your replies.
The material that I am going to simulate has properties that depend on coordinates in the reference configuration. 
I already thought of your suggestion. However, I do not know how I can implement it. 

I have already implemented such behavior in Abaqus UMAT using Fortran coding. Is there any way to have the Fortran code generated by mfront?

**Reply by thelfer** [reply]
You may want to read the `MFront` documentation about the `Abaqus` interface to understand how it works: https://thelfer.github.io/tfel/web/abaqus.html. The main idea is that we provide an `umat.cpp` file which loads shared libraries generated by `MFront` . 

The idea is to declare in the `MFront` source files, material properties or external state variables. Material properties   corresponds to the `PROPS` array of the `umat`, external state variables to the arrays `PRED` and `DPRED`. 

Let us consider this simple example:

```cxx
@DSL IsotropicPlasticMisesFlow;
@Behaviour Plasticity;

@MaterialProperty stress H;
@MaterialProperty stress s0;

@FlowRule {
  f = seq - H * p - s0;
  df_dseq = 1;
  df_dp = -H;
}
```

This behaviour declares four properties, as indicated by `mfront-query`:

```bash
- H
- s0
- YoungModulus (young): the Young's modulus of an isotropic material
- PoissonRatio (ν): the Poisson ratio of an isotropic material
```

This means that you can modify on the fly the first entry of the array PROPS using a function of the coordinates to compute the hardening slope `H`. There are then two solutions to do this:

- modify the sources generated by `MFront` for the behaviour that you want to use. In this case, calling `mfront --interface=abaqus Plasticity.mfront` generates a file named `umatPlasticity.cxx`. After modifying this file, you can compile the library using `mfront --obuild`. 
- modify the generic file `umat.cpp` that is passed to `Abaqus` (and compiled by `Abaqus`)

The first solution is cleaner, but more complex. The second is maybe more straigthforward.

**Reply by MojtabaFarrokh** [reply]
Thank you for your reply. I got your point. 
I think the first solution is better. However, I want to make sure of the procedure. Let's have a very basic example: a linear elastic material whose properties such as Young modulus and Poisson's ratio depend on the spatial coordinates. For example, E=f(X,Y,Z) and v=g(X,Y,Z). The mathematical forms of the function f and g are known and have some parameters (material constants) that must be defined and passed to the routine by the user. Could you please give me some hints about how I can implement it?

**Reply by thelfer** [reply]
@MojtabaFarrokh Let consider the following example: [abaqus.zip](https://github.com/thelfer/tfel/files/13485810/abaqus.zip)

The tar contains two files: `Plasticity.mfront` and `abaqusplasicity.mtest`

1. generate the sources:
    ```bash
    mfront --interface=abaqus Plasticity.mfront
    ```
2. Open `src/abaqusPlasticity.cxx` with you favorite editor (`emacs`, `vi`, `vscode`, ....)
    ```bash
    tfel-editor src/abaqusPlasticity.cxx &
    ```
3. Look at the PLASTICITY_3D and modify it as follows:
    ```
    double myPROPS[4] = {3e9, 200e6, PROPS[2], PROPS[3]};
    Plasticity_base_3D(STRESS,STATEV,DDSDDE,SSE,SPD,SCD,RPL,DDSDDT,DRPLDE,DRPLDT,
    STRAN,DSTRAN,TIME,DTIME,TEMP,DTEMP,PREDEF,DPRED,CMNAME,
    NDI,NSHR,NTENS,NSTATV,myPROPS,NPROPS,COORDS,DROT,PNEWDT,
    CELENT,DFGRD0,DFGRD1,NOEL,NPT,LAYER,KSPT,KSTEP,KINC,size);
    ```
4. Compile your shared library:
    ```bash
    mfront --obuild
    ```
5. Run the `MTest` example:
    ```bash
    mtest plasticity.mtest
    ```
    You will notice that the values in the `MTest` files are ignored and replaced by the values given in the `abaqusPlasticity.cxx` file, except for the Young's modulus and the Poisson's ratio
    ```
    Young's modulus: 1.5e+11
    Poisson's ratio: 0.3
    Yield strength: 2e+08
    Hardening slope: 3e+09
    iteration 3 : 1.80603e-19 5.13539e-08 (0.01 -0.00469935 -0.00469935 0 0 0)
    convergence, after 3 iterations, order 20.8243
    ```
 
You can now modify the example to define the `myPROPS` array using functions of the spatial coordinates.

The cumbersome part is that you have to modify the function for all supported hypotheses..

**Reply by MojtabaFarrokh** [reply]
Thank you very much.
I will try it.
